realm/realm-library/src/androidTest/java/io/realm/RealmQueryTest.java

```diff
@@ -2058,4 +2075,290 @@ public void distinct_invalidTypesLinkedFieldsThrows() {
+
+                    assertFalse(distinctBool.isLoaded());
+                    assertTrue(distinctBool.isValid());
+                    assertTrue(distinctBool.isEmpty());
+
+                    assertFalse(distinctLong.isLoaded());
+                    assertTrue(distinctLong.isValid());
+                    assertTrue(distinctLong.isEmpty());
+
+                    assertFalse(distinctDate.isLoaded());
+                    assertTrue(distinctDate.isValid());
+                    assertTrue(distinctDate.isEmpty());
+
+                    assertFalse(distinctString.isLoaded());
+                    assertTrue(distinctString.isValid());
+                    assertTrue(distinctString.isEmpty());
+
+                    Realm.asyncQueryExecutor.resume();
+
+                    distinctBool.addChangeListener(new RealmChangeListener() {
+                        @Override
+                        public void onChange() {
+                            assertEquals(2, distinctBool.size());
+                            signalCallbackFinished.countDown();
+                        }
+                    });
+
+                    distinctLong.addChangeListener(new RealmChangeListener() {
+                        @Override
+                        public void onChange() {
+                            assertEquals(numberOfBlocks, distinctLong.size());
+                            signalCallbackFinished.countDown();
+                        }
+                    });
+
+                    distinctDate.addChangeListener(new RealmChangeListener() {
+                        @Override
+                        public void onChange() {
+                            assertEquals(numberOfBlocks, distinctDate.size());
+                            signalCallbackFinished.countDown();
+                        }
+                    });
+
+                    distinctString.addChangeListener(new RealmChangeListener() {
+                        @Override
+                        public void onChange() {
+                            assertEquals(numberOfBlocks, distinctString.size());
+                            signalCallbackFinished.countDown();
+                        }
+                    });
+
+                    Looper.loop();
+                } catch (Throwable e) {
+                    e.printStackTrace();
+                    threadAssertionError[0] = e;
+
+                } finally {
+                    if (signalCallbackFinished.getCount() > 0) {
+                        signalCallbackFinished.countDown();
+                    }
+                    if (realm != null) {
+                        realm.close();
+                    }
+                    signalClosedRealm.countDown();
+                }
+            }
+        });
+
+        TestHelper.exitOrThrow(executorService, signalCallbackFinished, signalClosedRealm, backgroundLooper, threadAssertionError);
+    }
+
+    @Test
+    public void distinctAsync_withNull () throws Throwable {
```

Comment:
This and a lot of the other tests can be refactored to use `@RunTestInLooperThread` instead. It will allow you to remove a lot of the boilerplate and make the test safer. You can look in `RealmTests` for examples on how to use it.